File: assemblers/sic/java/SIC.java
// This is an SIC ASSEMBLER that creates the object code 
// file of a given SIC program. The pass 1 and pass 2 
// are also displayed in the output screen.

import java.util.*;
import java.io.*;

class SIC {
	Map<String, String> op = new HashMap<String, String>();
	List<String> ps_op = new ArrayList<String>();
	Map<String, String> SYMTAB = new LinkedHashMap<String, String>();
	
	int LC = 0;
	int start;
	int length;
	int end;
	String name = "";
	
	void initialise() {
		op.put("LDA", "00");
		op.put("LDX", "04");
		op.put("LDL", "08");
		op.put("STA", "0C");
		op.put("STX", "10");
		op.put("STL", "14");
		op.put("ADD", "18");
		op.put("SUB", "1C");
		op.put("MUL", "20");
		op.put("DIV", "24");
		op.put("COMP", "28");
		op.put("TIX", "2C");
		op.put("JEQ", "30");
		op.put("JGT", "34");
		op.put("JLT", "38");
		op.put("J", "3C");
		op.put("AND", "40");
		op.put("OR", "44");
		op.put("JSUB", "48");
		op.put("RSUB", "4C");
		op.put("LDCH", "50");
		op.put("STCH", "54");
		op.put("RD", "D8");
		op.put("WD", "DC");
		op.put("TD", "E0");
		op.put("STSW", "E8");
		
		ps_op.add("WORD");
		ps_op.add("BYTE");
		ps_op.add("RESW");
		ps_op.add("RESB");
	}
	
	void pass1(String filename) throws IOException {
		File file = new File(filename);
		FileWriter intFile = new FileWriter("intm_" + filename);
		Scanner scan = new Scanner(file);
		int lineNum = 1;
		
		displayPgm(filename);
		
		while(scan.hasNextLine()) {
			try {
				String str = scan.nextLine();
				String[] arr = str.split("\t");
				
				if(!arr[0].isEmpty()) {
					if(!SYMTAB.containsKey(arr[0]))
						SYMTAB.put(arr[0], Integer.toHexString(LC).toUpperCase());
					else {
						System.out.println("\n" + lineNum + ": Label \"" + arr[0] + "\" already used. Aborted.");
						intFile.close();
						scan.close();
						(new File("intm_" + filename)).delete();
						System.exit(0);
					}
				}
				
				if(arr[1].equals("START")) {
					if(arr.length == 3) {
						LC = Integer.parseInt(arr[2], 16);
						start = LC;
					}
					
					if(SYMTAB.containsKey(arr[0]))
						SYMTAB.put(arr[0], Integer.toHexString(LC).toUpperCase());
					
					name = arr[0];
				} else if(arr[1].equals("END")) {
					if(arr.length == 3) {
						try {
							end = Integer.parseInt(arr[2], 16);
						} catch (NumberFormatException e) {
							if(SYMTAB.containsKey(arr[2]))
								end = Integer.parseInt(SYMTAB.get(arr[2]), 16);
							else {
								System.out.println("\n" + lineNum + ": Label \"" + arr[2] + "\" not declared. Aborted.");
								intFile.close();
								scan.close();
								(new File("intm_" + filename)).delete();
								System.exit(0);
							}
						}
						
						intFile.write(Integer.toHexString(LC).toUpperCase() + "\t" + arr[1] + "\t" + arr[2] + "\n");
					} else {
						end = start;
						
						intFile.write(Integer.toHexString(LC).toUpperCase() + "\t" + arr[1] + "\n");
					}
					
					length = LC - start;
					break;
				} else if(op.containsKey(arr[1])) {
					if(arr.length == 3)
						intFile.write(Integer.toHexString(LC).toUpperCase() + "\t" + arr[1] + "\t" + arr[2] + "\n");
					else
						intFile.write(Integer.toHexString(LC).toUpperCase() + "\t" + arr[1] + "\n");
					
					LC += 3;
				} else if(ps_op.contains(arr[1])) {
					intFile.write(Integer.toHexString(LC).toUpperCase() + "\t" + arr[1] + "\t" + arr[2] + "\n");
					
					switch(arr[1]) {
						case "WORD":
							LC += 3;
							break;
						case "BYTE":
							if(arr[2].charAt(0) == 'C')
								LC += arr[2].length() - 3;
							else if(arr[2].charAt(0) == 'X')
								LC += (arr[2].length() - 3) / 2;
							break;
						case "RESW":
							LC += 3 * Integer.parseInt(arr[2]);
							break;
						case "RESB":
							LC += Integer.parseInt(arr[2]);
							break;
					}
				} else {
					System.out.println("\n" + lineNum + ": Invalid opcode \"" + arr[1] + "\". Aborted.");
					intFile.close();
					scan.close();
					(new File("intm_" + filename)).delete();
					System.exit(0);
				}
				
				lineNum++;
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
		
		intFile.close();
		
		displayPass1(filename);
		scan.close();
	}
	
	void pass2(String filename) throws IOException {
		File file = new File("intm_" + filename);
		FileWriter objFile = new FileWriter("obj_" + filename);
		FileWriter listFile = new FileWriter("list_" + filename);
		Scanner scan = new Scanner(file);
		int lineNum = 1;
		
		String startStr = Integer.toHexString(start).toUpperCase();
						
		int len = startStr.length();
		
		for(int i = 6; i > len; i--)
			startStr = "0" + startStr;
			
		String endStr = Integer.toHexString(end).toUpperCase();
						
		len = endStr.length();
		
		for(int i = 6; i > len; i--)
			endStr = "0" + endStr;
		
		String lengStr = Integer.toHexString(length).toUpperCase();
		
		len = lengStr.length();
		
		for(int i = 6; i > len; i--)
			lengStr = "0" + lengStr;
		
		for(int i = name.length(); i < 6; i++)
			name = name + " ";
		
		objFile.write("H^" + name + "^" + startStr + "^" + lengStr + "\n");
		
		int count = 0;
		String str1 = "";
		String startRec = "";
		
		while(scan.hasNextLine()) {
			try {
				String str = scan.nextLine();
				String[] arr = str.split("\t");
				
				String instr = "";
				
				if(op.containsKey(arr[1])) {
					if(arr.length == 3 && !SYMTAB.containsKey(arr[2].replace(",X", ""))) {
						System.out.println("\n" + lineNum + ": Label \"" + arr[2] + "\" not declared. Aborted.");
						objFile.close();
						listFile.close();
						scan.close();
						file.delete();
						(new File("list_" + filename)).delete();
						(new File("obj_" + filename)).delete();
						System.exit(0);
					} else if(arr.length == 3 && arr[2].contains(",X")) {
						instr = op.get(arr[1]);
						instr = instr.concat(
							Integer.toHexString(
								Integer.parseInt(
									SYMTAB.get(arr[2].replace(",X", "")), 16) + (int) Math.pow(2, 15)
							).toUpperCase()
						);
					} else {
						instr = op.get(arr[1]);
						instr = instr.equals("4C") ? instr.concat("0000") : instr.concat(SYMTAB.get(arr[2]));
					}
				}
				else {
					if(arr[1].equals("BYTE")) {
						if(arr[2].charAt(0) == 'X') {
							arr[2] = arr[2].substring(2, arr[2].length() - 1);
							instr = arr[2];
						}
						else if(arr[2].charAt(0) == 'C') {
							arr[2] = arr[2].substring(2, arr[2].length() - 1);
							
							char[] ch = arr[2].toCharArray();

							for (char c : ch)
								instr = instr.concat(String.format("%H", c));
						}
					} else if(arr[1].equals("WORD")) {
						instr = Integer.toHexString(Integer.parseInt(arr[2])).toUpperCase();
						
						len = instr.length();
						for(int i = 6; i > len; i--)
							instr = "0" + instr;
					}
				}
				
				listFile.write(arr[0] + "\t" + instr + "\n");
				
				lineNum++;
				
				if(instr.isEmpty() || count >= 30) {
					if(!str1.isEmpty()) {
						String recLen = Integer.toHexString(count).toUpperCase();
						if(recLen.length() == 1)
							recLen = "0" + recLen;
							
						objFile.write("T^"  + startRec + "^" + recLen + str1 + "\n");
						str1 = "";
						
						startRec = arr[0];
						
						len = startRec.length();
						for(int i = 6; i > len; i--)
							startRec = "0" + startRec;
					}
					
					count = 0;
				} 
				
				if(str1.isEmpty()) {
					startRec = arr[0];
					
					len = startRec.length();
					for(int i = 6; i > len; i--)
						startRec = "0" + startRec;
				}
				
				if(!instr.isEmpty()) {
					str1 = str1 + "^" + instr;
					count += instr.length() / 2;
				}
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
		
		objFile.write("E^" + endStr);
		
		listFile.close();
		objFile.close();
		
		displayPass2(filename);
		
		new File("list_" + filename).delete();
		file.delete();
		scan.close();
	}
	
	void displayPgm(String filename) throws IOException {
		int lineNum = 1;
		Scanner scan = new Scanner(new File(filename));
		
		System.out.println("\nYour SIC program is\n");
		while(scan.hasNextLine()) {
			System.out.print(lineNum);
			lineNum++;
			System.out.println(": " + scan.nextLine());
		}
		
		scan.close();
	}
	
	void displayPass1(String filename) throws IOException {
		int lineNum = 1;
		Scanner scan = new Scanner(new File("intm_" + filename));
		
		System.out.println("\nPass 1\n");
		while(scan.hasNextLine()) {
			System.out.print(lineNum);
			lineNum++;
			System.out.println(": " + scan.nextLine());
		}
		
		System.out.println("\nSymbol\tAddress");
		SYMTAB.forEach((key, value) ->
			System.out.println(key + "\t" + value));
		
		scan.close();
	}
	
	void displayPass2(String filename) throws IOException {
		Scanner scan = new Scanner(new File("list_" + filename));
		
		System.out.println("\nPass 2\n");
		
		while(scan.hasNextLine()) {
			System.out.println(scan.nextLine());
		}
		
		scan.close();
		
		scan = new Scanner(new File("obj_" + filename));
		
		System.out.println("\nObject code\n");
		
		while(scan.hasNextLine()) {
			System.out.println(scan.nextLine());
		}
		
		scan.close();
	}
	
	public static void main(String args[]) {
		SIC obj = new SIC();
		
		obj.initialise();
		
		System.out.print("Enter filename of SIC program: ");
		Scanner in = new Scanner(System.in);
		String filename = in.nextLine();
		
		try {
			obj.pass1(filename);
			obj.pass2(filename);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
}

File: assemblers/sic/c/sic.c
// This is an SIC ASSEMBLER that creates the object code 
// file of a given SIC program. The pass 1 and pass 2 
// are also displayed in the output screen.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

typedef struct map {
	char* key;
	char* value;
} map;

map op[26];
char* ps_op[4];
map SYMTAB[100];
int SYMTAB_SIZE = 0;
int LOCCTR = 0;
char name[6];
int start = 0;
int end = 0;
int length = 0;

int containsKey(map arr[], int size, char* key) {
	if(key == NULL)	
		return 0;
	
	for(int i = 0; i < size; i++)
		if(!strcmp(arr[i].key, key))
			return 1;
	
	return 0;
}

void updateKey(map arr[], int size, char* key, char* value) {
	if(key == NULL)
		return;
	
	for(int i = 0; i < size; i++)
		if(!strcmp(arr[i].key, key))
			arr[i].value = value;
}

char* getValue(map arr[], int size, char* key) {
	if(key == NULL)
		return NULL;
	
	for(int i = 0; i < size; i++)
		if(!strcmp(arr[i].key, key)) {
			char* temp = strdup(arr[i].value);
			return temp;
		}
	
	return NULL;
}

int containsElement(char* arr[], int size, char* key) {
	if(key == NULL)
		return 0;
	
	for(int i = 0; i < size; i++)
		if(!strcmp(arr[i], key))
			return 1;
	
	return 0;
}

char* intToHex(int num) {
	char* buff = malloc(7 * sizeof(char));
	sprintf(buff, "%X", num);
	return buff;
}

int hexToInt(char* buff) {
	int num;
	sscanf(buff, "%X", &num);
	return num;
}

char* stripX(char* token) {
	char* temp = strdup(token);
	
	for(int i = 0; i < strlen(token) - 1; i++) {
		if(temp[i] == ',' && temp[i+1] == 'X') {
			temp[i] = '\0';
			return temp;
		}
	}
	
	return temp;
}

void displayPgm(char* filename) {
	FILE* fp = fopen(filename, "r");
	char buff[64];
	
	printf("\nThe SIC program is given below:\n\n");
	while(1) {
		fgets(buff, 64, fp);
		if(feof(fp))
			break;
		fputs(buff, stdout);
	}
	
	fclose(fp);
}

void displayPass1() {
	FILE* fp = fopen("intm", "r");
	char buff[64];
	
	printf("\nPass 1:\n\n");
	while(1) {
		fgets(buff, 64, fp);
		if(feof(fp))
			break;
		fputs(buff, stdout);
	}
	
	fclose(fp);
	
	printf("\nSymbol Table:\n\n");
	
	for(int i = 0; i < SYMTAB_SIZE; i++)
		printf("%s\t%s\n", SYMTAB[i].key, SYMTAB[i].value);
}

void displayPass2() {
	FILE* fp = fopen("list", "r");
	char buff[64];
	
	printf("\nPass 2:\n\n");
	while(1) {
		fgets(buff, 64, fp);
		if(feof(fp))
			break;
		fputs(buff, stdout);
	}
	
	fclose(fp);
	
	fp = fopen("obj", "r");
	
	printf("\nObject code:\n\n");
	while(1) {
		fgets(buff, 64, fp);
		if(feof(fp))
			break;
		fputs(buff, stdout);
	}
	
	fclose(fp);
}

void pass1(char* filename) {
	displayPgm(filename);
	
	FILE* pgFile = fopen(filename, "r");
	FILE* intFile = fopen("intm", "w");
	char buff[64];
	
	while(1) {
		fgets(buff, 64, pgFile);
		
		if(feof(pgFile)) {
label:
			fclose(pgFile);
			fclose(intFile);
			displayPass1();
			return;
		}
		
		char* token = strtok(buff, "\t");
		if(token[strlen(token) - 1] == '\n')
			token[strlen(token) - 1] = '\0';
		
		int index = 0;
		char* temp = "";
		
		if(buff[0] == '\t')
			index = 1;
		else {
			temp = malloc(strlen(token) * sizeof(char));
			strcpy(temp, token);
		}
		
		while(token != NULL) {
			if(index == 0) {
				if(containsKey(SYMTAB, SYMTAB_SIZE, token)) {
					printf("\nLabel \"%s\" already used. Aborted.\n", token);
					fclose(pgFile);
					fclose(intFile);
					exit(0);
				}
				else {
					SYMTAB[SYMTAB_SIZE].key = malloc(strlen(token) * sizeof(char));
					strcpy(SYMTAB[SYMTAB_SIZE].key, token);
					SYMTAB[SYMTAB_SIZE].value = intToHex(LOCCTR);
					SYMTAB_SIZE++;
				}
			}
			else if(index == 1) {
				if(!strcmp(token, "START")) {
				
					token = strtok(NULL, "\t");
					if(token == NULL)
						break;
					else {
						LOCCTR = hexToInt(token);
						start = LOCCTR;
					}
					
					strcpy(name, temp);
					updateKey(SYMTAB, SYMTAB_SIZE, temp, intToHex(LOCCTR));
					
				} else if(!strcmp(token, "END")) {
				
					token = strtok(NULL, "\t");
					if(token == NULL) {
						end = start;
						break;
					}
					else {
						if(token[strlen(token) - 1] == '\n')
							token[strlen(token) - 1] = '\0';
						
						char* value = getValue(SYMTAB, SYMTAB_SIZE, token);
						
						if(value != NULL)
							end = hexToInt(value);
						else {
							printf("\nLabel \"%s\" undefined. Aborted.\n", token);
							fclose(pgFile);
							fclose(intFile);
							exit(0);
						}
					}
					
					length = LOCCTR - start;
					goto label;
					
				} else if(containsKey(op, 26, token)) {
				
					char* tempToken = malloc(strlen(token) * sizeof(char));
					
					strcpy(tempToken, token);
					
					token = strtok(NULL, "\t");
					
					if(token == NULL) {
						char str[20];
						strcpy(str, intToHex(LOCCTR));
						strcat(str, "\t");
						strcat(str, tempToken);
						strcat(str, "\n");
						fputs(str, intFile);
					} else {
						char str[20];
						strcpy(str, intToHex(LOCCTR));
						strcat(str, "\t");
						strcat(str, tempToken);
						strcat(str, "\t");
						strcat(str, token);
						fputs(str, intFile);
					}
					
					LOCCTR += 3;
					
				} else if(containsElement(ps_op, 4, token)) {
					
					char* tempToken = malloc(strlen(token) * sizeof(char));
					
					strcpy(tempToken, token);
					
					token = strtok(NULL, "\t");
					
					char str[20];
					strcpy(str, intToHex(LOCCTR));
					strcat(str, "\t");
					strcat(str, tempToken);
					strcat(str, "\t");
					strcat(str, token);
					fputs(str, intFile);
					
					if(!strcmp(tempToken, "WORD")) {
						LOCCTR += 3;
					} else if(!strcmp(tempToken, "BYTE")) {
						if(token[0] == 'C')
							LOCCTR += strlen(token) - 3;
						else if(token[0] == 'X')
							LOCCTR += (strlen(token) - 3) / 2;
					} else if(!strcmp(tempToken, "RESW")) {
						LOCCTR += 3 * atoi(token);
					} else if(!strcmp(tempToken, "RESB")) {
						LOCCTR += atoi(token);
					}
					
				} else {
					printf("\nInvalid opcode \"%s\". Aborted.", token);
					fclose(pgFile);
					fclose(intFile);
					exit(0);
				}
			}
			
			token = strtok(NULL, "\t");
			index++;
		}
	}
}

void pass2(char* filename) {
	FILE* intFile = fopen("intm", "r");
	FILE* listFile = fopen("list", "w");
	FILE* objFile = fopen("obj", "w");
	char buff[64];
	
	char* startStr = intToHex(start);
	
	int len = strlen(startStr);
	char temp[7] = "0";
	
	for(int i = 6; i > len; i--) {
		strcat(temp, startStr);
		strcpy(startStr, temp);
		strcpy(temp, "0");
	}
	
	char* endStr = intToHex(end);
	
	len = strlen(endStr);
	
	for(int i = 6; i > len; i--) {
		strcat(temp, endStr);
		strcpy(endStr, temp);
		strcpy(temp, "0");
	}
	
	char* lenStr = intToHex(length);
	
	len = strlen(lenStr);
	
	for(int i = 6; i > len; i--) {
		strcat(temp, lenStr);
		strcpy(lenStr, temp);
		strcpy(temp, "0");
	}
	
	for(int i = strlen(name); i < 6; i++)
		strcat(name, " ");
	
	int count = 0;
	char str[100] = "H^";
	char str1[100] = "";
	char startRec[7] = "";
	strcat(str, name);
	strcat(str, "^");
	strcat(str, startStr);
	strcat(str, "^");
	strcat(str, lenStr);
	strcat(str, "\n");
	fputs(str, objFile);
	
	while(1) {
		fgets(buff, 64, intFile);
		
		if(feof(intFile))
			break;
		
		char* token = strtok(buff, "\t");
		char* addr = strdup(token);
		
		token = strtok(NULL, "\t");
		if(token != NULL && token[strlen(token) - 1] == '\n')
			token[strlen(token) - 1] = '\0';
		
		char instr[15] = "";
		char* opcode = getValue(op, 26, token);
		
		char* tempToken = strdup(token);
		
		token = strtok(NULL, "\t");
		if(token != NULL && token[strlen(token) - 1] == '\n')
			token[strlen(token) - 1] = '\0';
		
		if(opcode != NULL) {
			strcat(instr, opcode);
			
			if(token == NULL) {
				strcat(instr, "0000");
			} else if(!containsKey(SYMTAB, SYMTAB_SIZE, stripX(token))) {
				printf("\nLabel \"%s\" not declared. Aborted.", token);
				fclose(objFile);
				fclose(listFile);
				fclose(intFile);
				exit(0);
			} else if(strstr(token, ",X")) {
				strcat(instr, intToHex(hexToInt(getValue(SYMTAB, SYMTAB_SIZE, stripX(token))) + pow(2, 15)));
			} else {
				strcat(instr, getValue(SYMTAB, SYMTAB_SIZE, token));
			}
		} else {
				if(!strcmp(tempToken, "BYTE")) {
					if(tempToken[0] == 'X') {
						for(int i = 2; i < strlen(token) - 1; i++) {
							instr[i-2] = token[i];
						}
					}
					else if(tempToken[0] == 'C') {
						for(int i = 2; i < strlen(token) - 1; i++) {
							char buff[2];
							sprintf(buff, "%X", token[i]);
							strcat(instr, buff);
						}
					}
				} else if(!strcmp(tempToken, "WORD")) {
					strcat(instr, intToHex(atoi(token)));
					
					len = strlen(instr);
					for(int i = 6; i > len; i--) {
						strcat(temp, instr);
						strcpy(instr, temp);
						strcpy(temp, "0");
					}
				}
		}
		
		strcpy(str, "");
		strcat(str, addr);
		strcat(str, "\t");
		strcat(str, instr);
		strcat(str, "\n");
		fputs(str, listFile);
		
		if(strlen(instr) == 0 || count >= 30) {
			if(strlen(str1) != 0) {
				char* recLen = intToHex(count);
				
				if(strlen(recLen) == 1) {
					strcat(temp, recLen);
					strcpy(recLen, temp);
					strcpy(temp, "0");
				}
				
				strcpy(str, "T^");
				strcat(str, startRec);
				strcat(str, "^");
				strcat(str, recLen);
				strcat(str, str1);
				strcat(str, "\n");
				fputs(str, objFile);
				
				strcpy(str1, "");
				count = 0;
				
				strcpy(startRec, addr);
				
				len = strlen(startRec);
				for(int i = 6; i > len; i--) {
					strcat(temp, startRec);
					strcpy(startRec, temp);
					strcpy(temp, "0");
				}
			}
		}
		
		if(strlen(startRec) == 0) {
			strcpy(startRec, addr);
			
			len = strlen(startRec);
			for(int i = 6; i > len; i--) {
				strcat(temp, startRec);
				strcpy(startRec, temp);
				strcpy(temp, "0");
			}
		}
		
		if(strlen(instr) != 0) {
			strcat(str1, "^");
			strcat(str1, instr);
			count += strlen(instr) / 2;
		}
	}
	
	strcpy(str, "E^");
	strcat(str, endStr);
	strcat(str, "\n");
	fputs(str, objFile);
	
	fclose(objFile);
	fclose(listFile);
	fclose(intFile);
	
	displayPass2();
}

void initialise() {
	op[0].key = "LDA";
	op[1].key = "LDX";
	op[2].key = "LDL";
	op[3].key = "STA";
	op[4].key = "STX";
	op[5].key = "STL";
	op[6].key = "ADD";
	op[7].key = "SUB";
	op[8].key = "MUL";
	op[9].key = "DIV";
	op[10].key = "COMP";
	op[11].key = "TIX";
	op[12].key = "JEQ";
	op[13].key = "JGT";
	op[14].key = "JLT";
	op[15].key = "J";
	op[16].key = "AND";
	op[17].key = "OR";
	op[18].key = "JSUB";
	op[19].key = "RSUB";
	op[20].key = "LDCH";
	op[21].key = "STCH";
	op[22].key = "RD";
	op[23].key = "WD";
	op[24].key = "TD";
	op[25].key = "STSW";
	
	op[0].value = "00";
	op[1].value = "04";
	op[2].value = "08";
	op[3].value = "0C";
	op[4].value = "10";
	op[5].value = "14";
	op[6].value = "18";
	op[7].value = "1C";
	op[8].value = "20";
	op[9].value = "24";
	op[10].value = "28";
	op[11].value = "2C";
	op[12].value = "30";
	op[13].value = "34";
	op[14].value = "38";
	op[15].value = "3C";
	op[16].value = "40";
	op[17].value = "44";
	op[18].value = "48";
	op[19].value = "4C";
	op[20].value = "50";
	op[21].value = "54";
	op[22].value = "D8";
	op[23].value = "DC";
	op[24].value = "E0";
	op[25].value = "E8";
	
	for(int i = 0; i < 4; i++)
		ps_op[i] = malloc(4 * sizeof(char)); 
	
	ps_op[0] = "WORD";
	ps_op[1] = "BYTE";
	ps_op[2] = "RESW";
	ps_op[3] = "RESB";
}

void main() {
	char* filename = malloc(20 * sizeof(char));
	
	initialise();
	
	printf("Enter filename of SIC program: ");
	fgets(filename, 20, stdin);
	
	for(int i = 0; i < 20; i++)
		if(filename[i] == '\n')
			filename[i] = '\0';
		
	pass1(filename);
	pass2(filename);
}
